### Preprocessing

In [249]:
# import statistical packages
import numpy as np
import pandas as pd

In [250]:
# import data visualisation packages
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

*I do not need to specify a separate 50% training dataset. Instead we use the [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) method from sklearn.*

In [251]:
from sklearn.model_selection import train_test_split

In [252]:
url = "/Users/arpanganguli/Documents/Professional/Finance/ISLR/Datasets/Auto.csv"
df = pd.read_csv(url)

In [253]:
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  18.0          8         307.0         130    3504          12.0    70   
1  15.0          8         350.0         165    3693          11.5    70   
2  18.0          8         318.0         150    3436          11.0    70   
3  16.0          8         304.0         150    3433          12.0    70   
4  17.0          8         302.0         140    3449          10.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [254]:
df.horsepower.dtype

dtype('int64')

*Quite annoyingly, I have to convert the datatype in horsepwer into float and store them in a separate column called 'hp'*

In [255]:
df['hp'] = df.horsepower.astype(float)

In [256]:
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  18.0          8         307.0         130    3504          12.0    70   
1  15.0          8         350.0         165    3693          11.5    70   
2  18.0          8         318.0         150    3436          11.0    70   
3  16.0          8         304.0         150    3433          12.0    70   
4  17.0          8         302.0         140    3449          10.5    70   

   origin                       name     hp  
0       1  chevrolet chevelle malibu  130.0  
1       1          buick skylark 320  165.0  
2       1         plymouth satellite  150.0  
3       1              amc rebel sst  150.0  
4       1                ford torino  140.0

In [257]:
df.hp.dtype

dtype('float64')

*Okay cool!*

### Regressions using random state = 1

**Simple Linear Regression**

In [258]:
X = df[['hp']]
y = df['mpg']

In [259]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=1)

In [260]:
X_train.shape

(198, 1)

In [261]:
y_train.shape

(198,)

In [262]:
X_test.shape

(199, 1)

In [263]:
y_test.shape

(199,)

In [264]:
df.shape

(397, 10)

*The Auto dataset contains 397 rows whereas the same dataset in the book example contains 392 rows. This can be explained
by the fact that some of the rows have missing values and have been deleted. I have, however, imputed those values. So,
I have the same number of rows as the original dataset. More information about imputation of missing values can be found 
[here](http://www.stat.columbia.edu/~gelman/arm/missing.pdf). In any case, it does not matter since the prime purpose of the chapter is to show relative differences in prediction abilities of different methodologies. So as long as the relative difference is more or less the same, the point still stands.*

In [265]:
from sklearn.linear_model import LinearRegression

In [266]:
lmfit = LinearRegression().fit(X_train, y_train)

In [267]:
lmpred = lmfit.predict(X_test)

In [268]:
from sklearn.metrics import mean_squared_error

In [269]:
MSE = mean_squared_error(y_test, lmpred)

In [270]:
round(MSE, 2)

24.23

**Polynomial Regression (horsepower$^2$)**

In [450]:
from sklearn.preprocessing import PolynomialFeatures as PF

In [451]:
X = df[['hp']]
X_ = pd.DataFrame(PF(2).fit_transform(X))
y = df[['mpg']]

In [452]:
X_.head()

0      1        2
0  1.0  130.0  16900.0
1  1.0  165.0  27225.0
2  1.0  150.0  22500.0
3  1.0  150.0  22500.0
4  1.0  140.0  19600.0

In [453]:
X_.drop(columns=0, inplace=True)

In [454]:
X_train, X_test, y_train, y_test = train_test_split(X_, y, test_size=0.5, random_state=1)

In [455]:
lmfit2 = LinearRegression().fit(X_train, y_train)

In [456]:
lmpred2 = lmfit2.predict(X_test)

In [457]:
MSE2 = mean_squared_error(y_test, lmpred2)

In [458]:
round(MSE2, 2)

18.87

**Polynomial Regression (horsepower$^3$)**

In [459]:
from sklearn.preprocessing import PolynomialFeatures as PF

In [460]:
X = df[['hp']]
X_ = pd.DataFrame(PF(3).fit_transform(X))
y = df[['mpg']]

In [461]:
X_.head()

0      1        2          3
0  1.0  130.0  16900.0  2197000.0
1  1.0  165.0  27225.0  4492125.0
2  1.0  150.0  22500.0  3375000.0
3  1.0  150.0  22500.0  3375000.0
4  1.0  140.0  19600.0  2744000.0

In [462]:
X_.drop(columns=0, inplace=True)

In [463]:
X_.head()

1        2          3
0  130.0  16900.0  2197000.0
1  165.0  27225.0  4492125.0
2  150.0  22500.0  3375000.0
3  150.0  22500.0  3375000.0
4  140.0  19600.0  2744000.0

In [464]:
X_train, X_test, y_train, y_test = train_test_split(X_, y, test_size=0.5, random_state=1)

In [465]:
lmfit3 = LinearRegression().fit(X_train, y_train)

In [466]:
lmpred3 = lmfit3.predict(X_test)

In [467]:
MSE3 = mean_squared_error(y_test, lmpred3)

In [468]:
round(MSE3, 2)

19.1

### Regressions using random state = 2

**Simple Linear Regression**

In [469]:
X = df[['hp']]
y = df['mpg']

In [470]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=2)

In [471]:
from sklearn.linear_model import LinearRegression

In [472]:
lmfit = LinearRegression().fit(X_train, y_train)

In [473]:
lmpred = lmfit.predict(X_test)

In [474]:
MSE = mean_squared_error(y_test, lmpred)

In [475]:
round(MSE, 2)

24.18

**Polynomial Regression (horsepower$^2$)**

In [506]:
from sklearn.preprocessing import PolynomialFeatures as PF

In [507]:
X = df[['hp']]
X_ = pd.DataFrame(PF(2).fit_transform(X))
y = df[['mpg']]

In [508]:
X_.head()

0      1        2
0  1.0  130.0  16900.0
1  1.0  165.0  27225.0
2  1.0  150.0  22500.0
3  1.0  150.0  22500.0
4  1.0  140.0  19600.0

In [509]:
X_.drop(columns=0, inplace=True)

In [510]:
X_.head()

1        2
0  130.0  16900.0
1  165.0  27225.0
2  150.0  22500.0
3  150.0  22500.0
4  140.0  19600.0

In [511]:
X_train, X_test, y_train, y_test = train_test_split(X_, y, test_size=0.5, random_state=2)

In [512]:
lmfit2 = LinearRegression().fit(X_train, y_train)

In [513]:
lmpred2 = lmfit2.predict(X_test)

In [514]:
MSE2 = mean_squared_error(y_test, lmpred2)

In [515]:
round(MSE2, 2)

19.05

**Polynomial Regression (horsepower$^3$)**

In [516]:
from sklearn.preprocessing import PolynomialFeatures as PF

In [517]:
X = df[['hp']]
X_ = pd.DataFrame(PF(3).fit_transform(X))
y = df[['mpg']]

In [518]:
X_.head()

0      1        2          3
0  1.0  130.0  16900.0  2197000.0
1  1.0  165.0  27225.0  4492125.0
2  1.0  150.0  22500.0  3375000.0
3  1.0  150.0  22500.0  3375000.0
4  1.0  140.0  19600.0  2744000.0

In [519]:
X_.drop(columns=0, inplace=True)

In [520]:
X_.head()

1        2          3
0  130.0  16900.0  2197000.0
1  165.0  27225.0  4492125.0
2  150.0  22500.0  3375000.0
3  150.0  22500.0  3375000.0
4  140.0  19600.0  2744000.0

In [521]:
X_train, X_test, y_train, y_test = train_test_split(X_, y, test_size=0.5, random_state=2)

In [522]:
lmfit3 = LinearRegression().fit(X_train, y_train)

In [523]:
lmpred3 = lmfit3.predict(X_test)

In [524]:
MSE3 = mean_squared_error(y_test, lmpred3)

In [525]:
round(MSE3, 2)

19.45

**Thus we see there is a difference in errors when we choose different training sets.**